src/GafferOSL/ShadingEngine.cpp

```diff
@@ -1241,6 +1508,222 @@ IECore::CompoundDataPtr ShadingEngine::shade( const IECore::CompoundData *points
+{
+	// Get the data for "P" - this determines the number of points to be shaded.
+
+	size_t numPoints = 0;
+
+	const OSL::Vec3 *p = nullptr;
+	if( const V3fVectorData *pData = points->member<V3fVectorData>( "P" ) )
+	{
+		numPoints = pData->readable().size();
+		p = reinterpret_cast<const OSL::Vec3 *>( &(pData->readable()[0]) );
+	}
+	else
+	{
+		throw Exception( "No P data" );
+	}
+
+	// Create ShaderGlobals, and fill it with any uniform values that have
+	// been provided.
+
+	// We don't ever pass this ShaderGlobals to OSL, but it works as a handy container
+	// for all the scalar values that we need to duplicate out onto each varying shader globals.
+	ShaderGlobals shaderGlobals;
+	memset( (void *)&shaderGlobals, 0, sizeof( ShaderGlobals ) );
+
+	const Gaffer::Context *context = Gaffer::Context::current();
+	shaderGlobals.time = context->getTime();
+
+	shaderGlobals.dPdx = uniformValue<V3f>( points, "dPdx" );
+	shaderGlobals.dPdy = uniformValue<V3f>( points, "dPdy" );
+	shaderGlobals.dPdz = uniformValue<V3f>( points, "dPdz" );
+
+	shaderGlobals.I = uniformValue<V3f>( points, "I" );
+	shaderGlobals.dIdx = uniformValue<V3f>( points, "dIdx" );
+	shaderGlobals.dIdy = uniformValue<V3f>( points, "dIdy" );
+
+	shaderGlobals.N = uniformValue<V3f>( points, "N" );
+	shaderGlobals.Ng = uniformValue<V3f>( points, "Ng" );
+
+	shaderGlobals.u = uniformValue<float>( points, "u" );
+	shaderGlobals.dudx = uniformValue<float>( points, "dudx" );
+	shaderGlobals.dudy = uniformValue<float>( points, "dudy" );
+
+	shaderGlobals.v = uniformValue<float>( points, "v" );
+	shaderGlobals.dvdx = uniformValue<float>( points, "dvdx" );
+	shaderGlobals.dvdy = uniformValue<float>( points, "dvdy" );
+
+	shaderGlobals.dPdu = uniformValue<V3f>( points, "dPdu" );
+	shaderGlobals.dPdv = uniformValue<V3f>( points, "dPdv" );
+
+	// Add a RenderState to the ShaderGlobals. This will
+	// get passed to our RendererServices queries.
+
+	RenderState renderState( points, transforms, contextVariablesNeeded, context );
+
+	// Get pointers to varying data, we'll use these to
+	// update the shaderGlobals as we iterate over our points.
+
+	const float *u = varyingValue<float>( points, "u" );
+	const float *v = varyingValue<float>( points, "v" );
+	const V2f *uv = varyingValue<V2f>( points, "uv" );
+	const V3f *n = varyingValue<V3f>( points, "N" );
+
+	/// \todo Get the other globals - match the uniform list
+
+	// Allocate data for the result
+
+	ShadingResults results( numPoints );
+
+	// Iterate over the input points, doing the shading as we go
+
+	tbb::enumerable_thread_specific<ThreadInfo> threadInfoCache;
+
+	const IECore::Canceller *canceller = context->canceller();
+
+
+	auto f = [&shadingSystem, &renderState, &results, &shaderGlobals, &p, &u, &v, &uv, &n, &shaderGroup, &threadInfoCache, canceller]( const tbb::blocked_range<size_t> &r )
+	{
+		ThreadInfo &threadInfo = threadInfoCache.local();
+		shadingSystem->batched<WidthT>().jit_group( &shaderGroup, threadInfo.shadingContext );
```

Comment:
I'm not sure if this was actually a problem, since I think the jitted form gets cached once it's been computed anyway ... but you're right that this looks weird and ugly.  Moved it outside the loop in 8c84600